### PR TITLE
Add UUID type and operators

### PR DESCRIPTION
**Pull request by Praveen2112**

### CODE REVIEW

**Comment by Praveen2112** [comment]
@pgagnon Thanks for pointing it out. Have removed it.

File: presto-main/src/test/java/io/prestosql/type/TestUUIDOperators.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import static io.prestosql.spi.type.VarcharType.VARCHAR;
+import static io.prestosql.type.UUIDType.UUID;
+
+public class TestUUIDOperators
+        extends AbstractTestFunctions
+{
+    @Test
+    public void testVarcharToUUIDCast()
+    {
+        assertFunction("CAST('1-2-3-4-1' AS UUID)", UUID, "00000001-0002-0003-0004-000000000001");
```
Comment (electrum) [inline presto-main/src/test/java/io/prestosql/type/TestUUIDOperators.java]: These shorter string representations don’t seem to be allowed by RFC 4122. I think the fact that Java accepts it is a bug caused by an implementation detail.

Let’s enforce the input length. Add tests for multiple UUID versions, plus the nil UUID.

Comment (dain) [reply presto-main/src/test/java/io/prestosql/type/TestUUIDOperators.java]: For the documentation, let's make sure we mention the RFC.

**Comment by dain** [comment]
@electrum any thoughts on what we do about the existing "uuid" function: https://github.com/prestosql/presto/blob/master/presto-accumulo/src/main/java/io/prestosql/plugin/accumulo/udf/AccumuloStringFunctions.java#L33-L39

File: presto-main/src/main/java/io/prestosql/type/UUIDOperators.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+import static io.prestosql.spi.function.OperatorType.HASH_CODE;
+import static io.prestosql.spi.function.OperatorType.INDETERMINATE;
+import static io.prestosql.spi.function.OperatorType.IS_DISTINCT_FROM;
+import static io.prestosql.spi.function.OperatorType.LESS_THAN;
+import static io.prestosql.spi.function.OperatorType.LESS_THAN_OR_EQUAL;
+import static io.prestosql.spi.function.OperatorType.NOT_EQUAL;
+import static io.prestosql.spi.function.OperatorType.XX_HASH_64;
+import static io.prestosql.type.UUIDType.UUID;
+
+public final class UUIDOperators
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/type/UUIDOperators.java]: For acronyms, we capitalize the first letter and use lower case for the rest when used in class or method names (`URL` -> `Url`, `HTTP` -> `Http`). So, in this case: `UuidOperators` .

**Comment by electrum** [comment]
I’m didn’t know that existed, and it’s not documented. We could replace it with a function that returns the new type.

**Comment by electrum** [comment]
@dain I’m didn’t know that existed, and it’s not documented. We could replace it with a function that returns the new type. PostgreSQL has more specifically named functions like `uuid_generate_v4`, but given that random is the most common type, it seems fine to use `uuid` over a more specific name like `uuid_random`, as that is the most common type (and likely the only one we will ever need to support).

File: presto-main/src/main/java/io/prestosql/type/UUIDOperators.java (hunk 2 above)
Comment (Praveen2112) [reply presto-main/src/main/java/io/prestosql/type/UUIDOperators.java]: Moved to the above format

File: presto-main/src/test/java/io/prestosql/type/TestUUIDOperators.java (hunk 1 above)
Comment (Praveen2112) [reply presto-main/src/test/java/io/prestosql/type/TestUUIDOperators.java]: Have added this to types documentation and also handled for various version of UUID

File: presto-main/src/test/java/io/prestosql/type/TestUUIDType.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+{
+    public TestUUIDType()
+    {
+        super(UUID, String.class, createTestBlock());
+    }
+
+    public static Block createTestBlock()
+    {
+        BlockBuilder blockBuilder = UUID.createBlockBuilder(null, 1);
+        UUID.writeSlice(blockBuilder, getSliceForUUID("6b5f5b65-67e4-43b0-8ee3-586cd49f58a0"));
```
Comment (electrum) [inline presto-main/src/test/java/io/prestosql/type/TestUUIDType.java]: This could be simplified a bit:
```java
for (int i = 0; i < 10; i++) {
    UUID.writeSlice(Slices.wrappedLongArray(0x6B5F5B6567E443B0, 0x8EE3586CD49F58A0 + i);
}
```
This also makes it clear that the value is changing (you need to look hard for the last character)

Or if you think it's better with text:
```java
for (int i = 0; i < 10; i++) {
    String uuid = "6b5f5b65-67e4-43b0-8ee3-586cd49f58a" + i
    UUID.writeSlice(castFromVarcharToUuid(utf8slice(uuid)));
}
```

File: presto-main/src/test/java/io/prestosql/type/TestUUIDType.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    @Override
+    protected Object getNonNullValue()
+    {
+        return Slices.wrappedLongArray(0, 0);
+    }
+
+    @Test
+    public void testDisplayName()
+    {
+        assertEquals((UUID).getDisplayName(), "uuid");
```
Comment (electrum) [inline presto-main/src/test/java/io/prestosql/type/TestUUIDType.java]: No need for parens

File: presto-main/src/main/java/io/prestosql/type/UuidOperators.java [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    @SqlType(StandardTypes.BIGINT)
+    public static long xxHash64(@SqlType(StandardTypes.UUID) Slice value)
+    {
+        return XxHash64.hash(value);
+    }
+
+    @LiteralParameters("x")
+    @ScalarOperator(CAST)
+    @SqlType(StandardTypes.UUID)
+    public static Slice castFromVarcharToUUID(@SqlType("varchar(x)") Slice slice)
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/type/UuidOperators.java]: Rename to `castFromVarcharToUuid`. Same for the rest of the methods in this class.

File: presto-main/src/test/java/io/prestosql/type/TestUuidOperators.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        assertOperator(HASH_CODE, "CAST(null AS UUID)", BIGINT, null);
+        assertOperator(HASH_CODE, "UUID '12151fd2-7586-11e9-8f9e-2a86e4085a59'", BIGINT, hashFromType("12151fd2-7586-11e9-8f9e-2a86e4085a59"));
+    }
+
+    private static long hashFromType(String uuidString)
+    {
+        BlockBuilder blockBuilder = UUID.createBlockBuilder(null, 1);
+        java.util.UUID uuid = java.util.UUID.fromString(uuidString);
+        UUID.writeSlice(blockBuilder, Slices.wrappedLongArray(uuid.getMostSignificantBits(), uuid.getLeastSignificantBits()));
```
Comment (electrum) [inline presto-main/src/test/java/io/prestosql/type/TestUuidOperators.java]:
```java
UUID.writeSlice(blockBuilder, castFromVarcharToUuid(utf8Slice(uuidString)));

File: presto-docs/src/main/sphinx/language/types.rst [hunk 7]
```diff
@@ -237,6 +237,19 @@ Network Address
...
 
+UUID
+----
+
+.. _uuid_type:
+
+``UUID``
+^^^^^^^^
+
+    This type represents Universally Unique Identifier. The format defined
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply presto-docs/src/main/sphinx/language/types.rst]: done

**Comment by Praveen2112** [comment]
@electrum Have handled all the commits. Please let me know if there are any more changes.

**Comment by electrum** [comment]
Merged, thanks!
